string_manipulation.rb

```diff
@@ -1,30 +1,104 @@
 # A method to reverse a string in place.
 def string_reverse(my_string)
-  puts "NOT IMPLEMENTED"
+  # swap each element
+  # move an index
+  i = 0
+  n = my_string.length
+  while i < n/2
+    first = my_string[i]
+    my_string[i] = my_string[n-i-1]
+    my_string[n-i-1] = first
+    i += 1
+  end
 end
 
 # A method to reverse each word in a sentence, in place.
 def reverse_words(my_words)
-  puts "NOT IMPLEMENTED"
+  # iterate through each characters
+  # if space leave alone
+  # create stack of letters
+  # pop off onto string
+  n = my_words.length
+  i = 0
+  # CAN_ 0123 NAC
+  while i < n
+    word = []
+    j = i
+    while my_words[j] != " " && j < n
+      word << my_words[j]
+      j += 1
+    end
+    while i < j
+      my_words[i] = word.pop
+      i += 1
+    end
+    i += 1
+  end
 end
 
 # A method to reverse the words in a sentence, in place.
 def reverse_sentence(my_sentence)
-  puts "NOT IMPLEMENTED"
+  string_reverse(my_sentence)
+  reverse_words(my_sentence)
 end
 
 # A method to check if the input string is a palindrome.
 # Return true if the string is a palindrome. Return false otherwise.
 def palindrome_check(my_phrase)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  j = my_phrase.length - 1
+  while i < j
+    if my_phrase[i] == " "
+      i += 1
+    end
+    if my_phrase[j] == " "
+      j -= 1
+    end
+    if my_phrase[i].downcase != my_phrase[j].downcase
+      return false
+    else
+      i += 1
+      j -= 1
+    end
+  end
   return true
 end
 
 # A method that updates the string by replacing consecutive repeating characters
 # with a number representing the frequncy. The replacement is done only if the
 # string length will get reduced by the process.
+# 012345678901
+# aaabbbbbcccc will get replace to "a3b5c4"
+# a3abbbbbcccc
+# 012345678901
+# a3bbbbbcccc-
+# a3b5bbbcccc-
+# a3b5cccc----
 def encode_repeating(my_string)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  n = my_string.length
+  while i < n
+    first = my_string[i]
+    i += 1
+    j = 1
+    while my_string[i] == first
+      j += 1
+      i += 1
+    end
+    if j > 2
+      k = i-j+1
+      my_string[k] = j.to_s
+      k += 1
+      shift = j-2

```

Comment:
Nice work implementing this method. :-) Now we can discuss optimizations.
Instead of doing this copy remaining and slice! on each letter encode, you could do it just once after the loop ends. To make this happen, you'll need to keep separate track of where you are reading and counting, verses where you are entering/copying. If you can figure this out, it will make it a O(n) time complexity.